### PR TITLE
More robust check for open ssh port in SUT

### PR DESCRIPTION
**Pull request by Soulofdestiny**
[related poo](https://progress.opensuse.org/issues/33202)
[MR for adding package to the worker](https://gitlab.suse.de/openqa/salt-states-openqa/merge_requests/49)

verification in working case after 4 retries:
verification in working case after 4 retries

failing case with proper die message:
failing case with proper die message

@coolo @okurz

### CODE REVIEW

File: lib/opensusebasetest.pm [hunk 1]
```diff
@@ -334,8 +335,22 @@ sub wait_boot {
...
+                die "The SSH Port in the SUT could not be reached within 1 minute, considering a product issue" if $i == $retries;
+                my $port       = 22;
+                my %ports      = (tcp => {$port => {},});
+                my $host_check = check_ports($virsh_guest, 5, \%ports);
+                if ($host_check->{tcp}{$port}{open} == 1) {
+                    record_info("ssh port open", "check for port $port on $virsh_guest successful");
+                    last;
+                }
+                else {
+                    record_info("ssh port closed", "check for port $port on $virsh_guest failed", result => 'fail');
```
Comment (rwx788) [inline lib/opensusebasetest.pm]: Don't we want to retry 12 times before failing here? As with `result => 'fail'` we will exit after first failure. Or does it work differently?

Comment (Soulofdestiny) [reply lib/opensusebasetest.pm]: works differently - `result => 'fail'` just draws a red border for record_info - though I'm not convinced if we want to have that

Comment (asmorodskyi) [reply lib/opensusebasetest.pm]: and I actually think you should change this . Let's keep red color for real failures . When it is expected failure you shouldn't mix it with unexpected ones .

File: cpanfile [hunk 2]
```diff
@@ -4,6 +4,7 @@ requires 'XML::LibXML';
 requires 'XML::Writer';
 requires 'XML::Simple';
 requires 'IO::File';
+requires 'IO::Socket::PortState';
```
Comment (coolo) [inline cpanfile]: remember that workers don't care about cpanfile. If you merge this you will break basically all tests on o3 *and* osd

Comment (okurz) [reply cpanfile]: Hence the salt recipe change. We can think about running isotovideo in a container where we install with cpan :)

**Comment by okurz** [comment]
@asmoroskyi is on it ;P

**Comment by asmorodskyi** [comment]
I sign off from this task

**Comment by okurz** [comment]
@Soulofdestiny so it seems we have the following options how to go forward:

* **O1:** Full saltify o3 so that we can also ensure packages are installed in the same way as for osd
* **O2:** Because it's a mess anyway, just make sure the package is installed manually on all o3 workers
* **O3:** Adapt os-autoinst to install packages from the tests distri on startup of tests using e.g. cpan with custom lib or container
* **O4:** Go back to the old way with less dependencies
* **O5:** Move the check into the backend along with the additional package dependency to cpanfile as well as ensuring that the devel:openQA package in OBS specifies this requirement

I guess you want to go with O5 or O4, maybe O3

**Comment by Soulofdestiny** [comment]
trying O6 which santi proposed, using IO::Socket::INET, which is already a dependency of os-autoinst,
otherwise going back to the old way for now

**Comment by Soulofdestiny** [comment]
so it works with `IO::Socket::INET` and looks like this:
selection_004

**Comment by Soulofdestiny** [comment]
I guess we can with the flag for record_info

**Comment by okurz** [comment]
Should be neither fail nor softfail as there is nothing wrong when ssh can not reach the target port when it is simply not reachable yet. Blue is the default color. Did you try to not set any flag at all or is there a specific reason why you set it?

**Comment by Soulofdestiny** [comment]
> Did you try to not set any flag at all or is there a specific reason why you set it?

Yes - it defaults to 'ok' which gives a green border - which is imo also wrong and misleading to the reviewers

unfortunately there are only three options allowed as you can see in https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L185

**Comment by Soulofdestiny** [comment]
given the fact the we also have the red border for the YaST log info screen
and the fact that red is expressing the most attention to the reviewer I'd still go for the red border

**Comment by rwx788** [comment]
I'm with @Soulofdestiny here, as in such case something actually went wrong, hence failed. With other borders we tend to hide issues.

**Comment by okurz** [comment]
@rwx788 I think you haven't read my comments above. The red borders in y2log parsing are also not nice but actually there are bugs about it already. I would go for showing something as red really only in the case of something failed. Trying to access a machine over ssh when it's not yet fully booted should not be regarded as failure

**Comment by rwx788** [comment]
@okurz I've read your comment, so don't be grumpy. You've shared your opinion, that's good, but that's not the only opinion which may exist. As you can see in the code, we check if sshd is running before retrying, so it's not expected that we fail to connect straight away. Therefore it's showing the issue, but doesn't fail the test. If you find it wrong, create PR which changes both cases mentioned above and if it gets merged - it gets merged. I find it wrong to waste days on these discussions without having functional part done.

**Pull request merged**